docs(RTL): add RTL handbook (#3745)
patternfly/patternfly-org: packages/documentation-site/patternfly-docs/content/developer-resources/rtl-handbook.md

---
id: Right-to-left handbook

> title: PatternFly right-to-left handbook

Reviewer: While the main focus is RTL, we do touch on stuff not specific to RTL on this page; wondering if something like "Bidrection(ality) handbook" would be a more fitting title?

Author: While I agree that it's more technically correct, I also think that most people would be more likely to find the information if we call it right-to-left - wdyt, @edonehoo? 

On this and the other wording changes - I defer to @edonehoo and if you want to merge and then Erin you add/change whatever you'd like, that's totally fine with me - I really just wanted to get the technical details in and get this documentation started.

Reviewer: Yeah that makes sense regarding the naming. I'm good with it as-is.

And I'd be good with a followup to tweak things further if need be; this is definitely in a good spot to get up with the rest of the RTL updates

Reviewer 2: I think I left that as an idea somewhere within my many edits - I also thought "bidirectionality handbook" was more accurate. Or "bidirectional language handbook/support"? But I also agree that it's more nitpicky and definitely not blocking getting this information live

section: developer-resources

---

To allow for internationalization of your product's content, it is important to implement bidirectional language support by developing for both right-to-left (RTL) and left-to-right (LTR) languages. 

> The default browser behavior displays content following LTR formatting. To support RTL languages, like Arabic and Hebrew, you will need to explicitly add RTL support. To support RTL languages the browser uses `dir="RTL"` and `lang="{language code}"` to lay out the page appropriately for the selected language. Setting `dir="RTL"` causes the browser to reverse and lay out the contents from right to left. The W3C provides [information about internationalization and the bidirectional algorithm](https://www.w3.org/International/techniques/authoring-html#direction).

Reviewer 2: ```suggestion
The default browser behavior displays content following LTR formatting. To support RTL languages, like Arabic and Hebrew, you will need to explicitly add RTL support. In order to support RTL languages, the browser uses `dir="RTL"` and `lang="{language code}"` to lay out the page appropriately for the selected language. Setting `dir="RTL"` causes the browser to reverse and lay out the contents from right to left. The W3C provides [information about internationalization and the bidirectional algorithm](https://www.w3.org/International/techniques/authoring-html#direction).
```

The world's smallest edit

## Support in PatternFly

PatternFly components have been written to use logical properties - `block-start`, `inline-end`, `block-end`, and `inline-start` - rather than `top`, `right`, `bottom`, and `left`. This allows styles like margin and padding to be correctly laid out in both LTR and RTL. However, to avoid a sweeping breaking change, PatternFly variable names still use `top`, `bottom`, `left`, and `right`.

> PatternFly components have been written to reverse directional icons that are part of the component - for example, the caret indicating expansion of various components is properly reversed. 

Reviewer: Not a blocker, but what do you think about adding a couple of images to show an LTR/RTL difference? "In the following images of our Component, the caret icon is facing towards/away from the toggle text when the toggle is collapsed in both LTR and RTL."

> Other icons you may have in your interface will not be mirrored automatically, as it may be appropriate to mirror some but not others. Be sure to review all icons in your product and adjust appropriately. PatternFly has modifier classes listed below that can help with that.
> <br />
> 
> ## PatternFly modifiers supporting RTL

Reviewer: Not a blocker, but would this paragraph make more sense in this section with some tweaking?

```suggestion
<br />

## PatternFly modifiers supporting RTL

Some elements you have in your interface will not be adjusted for RTL by PatternFly automatically, or you may require more customized adjustment. The following PatternFly modifier classes can be used for more fine-tuned control.
```

Reviewer 2: I like this phrasing!

Use the modifier class `.pf-v5-m-mirror-inline-rtl` to reverse an icon or other element only when in RTL.

To set direction manually on an element, use `.pf-v5-m-dir-ltr` or `.pf-v5-m-dir-rtl`.

> To reverse a value, you can multiply by the variable `--#{$pf-global}--inverse--multiplier`, which will have a value of 1 in LTR and -1 in RTL.

Reviewer: ```suggestion
To reverse a value, you can multiply by the variable `--pf-v5-global--inverse--multiplier`, which will have a value of 1 in LTR and -1 in RTL.
```

> To reverse a value, you can multiply by the variable `--pf-v5-global--inverse--multiplier`, which will have a value of 1 in LTR and -1 in RTL.

Reviewer 3: ```suggestion
To reverse a value...
```

I wonder if we should say "to get the inverse of a numerical value..." or instead of "numerical", "number, length, or percentage" which is more specific to the [value types](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#numbers_lengths_and_percentages) it will work with.

## Support in your product

Care should be taken to handle icons appropriately for RTL. [Sometimes an icon should be reversed in RTL](https://m2.material.io/design/usability/bidirectionality.html#mirroring-elements) (for example, when indicating a direction such as an expansion caret), and sometimes it should not be reversed (such as a logo, or an icon indicating time). When the icon should be reversed, the modifier `.pf-v5-m-mirror-inline-rtl` can be used.

## Best practices and resources
- There are [many nuances to how content is laid out through LTR and RTL](https://ltr.wtf/explained/bidiintro.html#the-good-the-bad-and-the-ugly), so it isn't always obvious what the correct result should be. Testing your site with native speakers is always recommended. 
- Material Design summarizes [which UI elements should be reversed in an RTL layout](https://m2.material.io/design/usability/bidirectionality.html#mirroring-layout).

> - For a comprehensive discussion on RTL styling, refer to the [RTL Styling 101](https://rtlstyling.com/posts/rtl-styling) guide.

Reviewer 3: The w3c has a couple of good links, too

https://www.w3.org/International/articles/inline-bidi-markup/
https://www.w3.org/International/questions/qa-html-dir

---

> id: Right to left handbook

Reviewer 2: ```suggestion
id: Right-to-left handbook
```
Or should this be called "Bidirectionality handbook"? Not sure which term is most recognizable to devs

> title: PatternFly Right to Left (RTL) handbook

Reviewer 2: ```suggestion
title: PatternFly right-to-left handbook
```

Would recommend avoiding acronyms in headings!

Author: I don't think I've ever sees it hyphenated but now I'm doubting myself lol. I made some updates hopefully incorporating all these comments - will revisit tomorrow to make sure I got everything!

Reviewer 2: I'll defer to what you decide is correct since it's not a concept I'm personally familiar with! I saw it once on Material and also on Apple: https://developer.apple.com/design/human-interface-guidelines/right-to-left, but I also see it without hyphens across some websites when I Google 🧐.

Material Design summarizes [what should be reversed in an RTL layout](https://m2.material.io/design/usability/bidirectionality.html#mirroring-layout).

For a comprehensive discussion on RTL Styling, the [RTL Styling 101](https://rtlstyling.com/posts/rtl-styling) guide is very useful.

## Support in PatternFly

PatternFly components have been written to use logical properties - `inline-start`, `inline-end`, `block-start`, and `block-end` - rather than `top`, `bottom`, `left`, and `right` so that attributes like margin and padding will be correctly reflected when laid out in RTL. However, to avoid a sweeping breaking change, PatternFly variable names still use `top`, `bottom`, `left`, and `right`.

PatternFly components have been written to reverse directional icons that are part of the component - for example, the caret indicating expansion of various components is properly reversed.  Likewise, any components using transforms for positioning have also been handled by reversing the direction of the `translateX` value.

> Other icons you may have in your interface will not be mirrored automatically, as it may be appropriate to mirror some but not others.

Reviewer 3: Maybe add something like "Be sure to review all icons in your product and adjust appropriately. PatternFly has modifier classes listed below that can help with that."?

<br />

## Modifiers and mixins supporting RTL

> Care should be taken to handle icons appropriately for RTL. [Sometimes an icon should be reversed in RTL](https://m2.material.io/design/usability/bidirectionality.html#mirroring-elements) (for example, when indicating a direction such as an expansion caret), and sometimes it should not be reversed (such as a logo, or an icon indicating time). When the icon should be reversed, the modifier `.pf-v5-m-mirror-inline-rtl` can be used.

Reviewer 3: I wonder if this section should just be a bullet list or something of our modifiers, with a description of what they do and when to use them. And maybe the rest of the text in this section should be in another section for things products should do to better support custom/user-supplied stuff between LTR/RTL?

Reviewer 2: this sounds like a good idea if it's important to have a good/reliable reference for modifiers ^. & then like Michael said - a "best practice" H2 section?

Using logical properties handles the majority of cases - but using the `translateX` function can still present a problem. In this case, we've used a mixin that allows us to provide a value for LTR and a value for RTL. In most cases, the RTL value uses another Sass function to inverse (multiply by -1) the value so that the translate function shifts the element in the correct direction.
```
@include pf-v5-bidirectional-style(
  $prop: transform,
  $ltr-val: translateX(var(--translation-value)),
  $rtl-val: translateX(#{pf-v5-calc-inverse(var(--translation-value))})
); 

> ```

Reviewer 3: I'm not sure if we should include this... I doubt most folks are compiling our SCSS, so I'm not sure what value this would add. If we wanted to, we could mention the `--#{$pf-global}--inverse--multiplier` var that is `1` by default and `-1` in RTL?
